generics, cgen: invalid cgen when mixing type aliases

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
|V full version      |V 0.4.9 5b44b67.af1ef92
|:-------------------|:-------------------
|OS                  |macos, macOS, 15.2, 24C101
|Processor           |10 cpus, 64bit, little endian, Apple M4
|Memory              |0.4GB/16GB
|                    |
|V executable        |/Users/leodev/p/v/v/v
|V last modified time|2025-01-05 15:18:43
|                    |
|V home dir          |OK, value: /Users/leodev/p/v/v
|VMODULES            |OK, value: /Users/leodev/.vmodules
|VTMP                |OK, value: /tmp/v_501
|Current working dir |OK, value: /Users/leodev/p/v/fast2
|                    |
|Git version         |git version 2.47.0
|V git status        |weekly.2024.53-31-gaf1ef929 (1 commit(s) behind V master)
|.git/config present |true
|                    |
|cc version          |Apple clang version 16.0.0 (clang-1600.0.26.6)
|gcc version         |Apple clang version 16.0.0 (clang-1600.0.26.6)
|clang version       |Apple clang version 16.0.0 (clang-1600.0.26.6)
|tcc version         |tcc version 0.9.28rc 2024-02-05 HEAD@105d70f7 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 713692d4
|emcc version        |N/A
|glibc version       |N/A
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  repro.v && repro`

```v
type I64 = i64

fn test[T](a ?T) ? {
	return none
}

fn main() {
	a := ?i64(123)
	b := test[I64](a)
	println(b)
}
```

**What did you see?**

```
================== C compilation error (from cc): ==============
cc: /tmp/v_501/repro.01JGVHF8TVE1GJF24R0361RXXP.tmp.c:5871:29: error: initializing 'main__I64' (aka 'long long') with an expression of incompatible type '_option_i64' (aka 'struct _option_i64')
cc:  5871 |         _option_ok(&(main__I64[]) {a }, (_option*)(&_t2), sizeof(main__I64));
cc:       |                                    ^
cc: 1 error generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Successful compilaton

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzdhYTYxNmYxNDRmNTg1YWU1MjkyZGQiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0._AX10U_hCly-MZp5yyjIyOuR0Sjfv_PyIR92-DMY1V0">Huly&reg;: <b>V_0.6-21813</b></a></sub>

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -55,7 +55,7 @@ fn (mut g Gen) expr_with_opt_or_block(expr ast.Expr, expr_typ ast.Type, var_expr
 }
 
 // expr_opt_with_alias handles conversion from different option alias type name
-fn (mut g Gen) expr_opt_with_alias(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type, tmp_var string) string {
+fn (mut g Gen) expr_opt_with_alias(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type) string {
 	styp := g.base_type(ret_typ)
 
 	line := g.go_before_last_stmt().trim_space()
@@ -63,21 +63,23 @@ fn (mut g Gen) expr_opt_with_alias(expr ast.Expr, expr_typ ast.Type, ret_typ ast
 
 	ret_var := g.new_tmp_var()
 	ret_styp := g.styp(ret_typ).replace('*', '_ptr')
-	g.writeln('${ret_styp} ${ret_var} = {0};')
+	g.writeln('${ret_styp} ${ret_var} = {.state=2, .err=_const_none__, .data={EMPTY_STRUCT_INITIALIZATION}};')
 
-	g.write('_option_clone((${option_name}*)')
-	has_addr := expr !in [ast.Ident, ast.SelectorExpr]
-	if has_addr {
-		expr_styp := g.styp(expr_typ).replace('*', '_ptr')
-		g.write('ADDR(${expr_styp}, ')
-	} else {
-		g.write('&')
+	if expr !is ast.None {
+		g.write('_option_clone((${option_name}*)')
+		has_addr := expr !in [ast.Ident, ast.SelectorExpr]
+		if has_addr {
+			expr_styp := g.styp(expr_typ).replace('*', '_ptr')
+			g.write('ADDR(${expr_styp}, ')
+		} else {
+			g.write('&')
+		}
+		g.expr(expr)
+		if has_addr {
+			g.write(')')
+		}
+		g.writeln(', (${option_name}*)&${ret_var}, sizeof(${styp}));')
 	}
-	g.expr(expr)
-	if has_addr {
-		g.write(')')
-	}
-	g.writeln(', (${option_name}*)&${ret_var}, sizeof(${styp}));')
 	g.write(line)
 	if g.inside_return {
 		g.write(' ')
@@ -97,7 +99,7 @@ fn (mut g Gen) expr_opt_with_cast(expr ast.Expr, expr_typ ast.Type, ret_typ ast.
 		return g.expr_with_opt(expr, expr_typ, ret_typ)
 	} else {
 		if expr is ast.CallExpr && expr.return_type.has_flag(.option) {
-			return g.expr_opt_with_alias(expr, expr_typ, ret_typ, '')
+			return g.expr_opt_with_alias(expr, expr_typ, ret_typ)
 		} else {
 			past := g.past_tmp_var_new()
 			defer {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5223,7 +5223,7 @@ fn (mut g Gen) cast_expr(node ast.CastExpr) {
 					g.writeln('_option_ok(&(${g.styp(parent_type)}[]) { ${tmp_var2} }, (${option_name}*)&${tmp_var}, sizeof(${g.styp(parent_type)}));')
 					g.write2(cur_stmt, tmp_var)
 				} else if node.expr_type.has_flag(.option) {
-					g.expr_opt_with_cast(node.expr, expr_type, node.typ)
+					g.expr_opt_with_alias(node.expr, expr_type, node.typ)
 				} else {
 					g.expr_with_opt(node.expr, expr_type, node.typ)
 				}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2771,8 +2771,8 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 		g.write('->val')
 		return
 	} else if expected_type.has_flag(.option) {
-		if arg_sym.info is ast.Alias && expected_type != arg_typ {
-			g.expr_opt_with_alias(arg.expr, arg_typ, expected_type, '')
+		if (arg_sym.info is ast.Alias || exp_sym.info is ast.Alias) && expected_type != arg_typ {
+			g.expr_opt_with_alias(arg.expr, arg_typ, expected_type)
 		} else {
 			g.expr_with_opt(arg.expr, arg_typ, expected_type)
 		}

--- a/vlib/v/tests/options/option_generic_alias_test.v
+++ b/vlib/v/tests/options/option_generic_alias_test.v
@@ -1,0 +1,17 @@
+type I64 = i64
+
+fn test[T](a ?T) ?T {
+	w := ?T(a)
+	return w
+}
+
+fn test_main() {
+	a := ?i64(123)
+	b := test[I64](a)
+	println(b)
+	assert b != none
+	assert b? == 123
+	c := test[I64](none)
+	println(c)
+	assert c == none
+}

